File: Olympiad/COI/official/2007/policija.cpp
/*
  Croatian Olympiad in Informatics 2007
  Task POLICIJA
*/

#include <algorithm>
#include <cstdio>
#include <vector>

using namespace std;

int n, m;
struct edge {
   int u, v;
   edge( int U, int V ) { u = U; v = V; }
};
bool operator < ( const edge &A, const edge &B ) { return A.u < B.u; }

struct sparse_graph {
   vector<edge> E;
   vector< vector<edge>::iterator > V;

   void insert_edge( const edge &e ) {
      E.push_back( e );
   }

   void init() {
      V.resize(n+1);
      sort( E.begin(), E.end() );
      V[0] = E.begin();
      for( int i = 1; i <= n; ++i )
         for( V[i] = V[i-1]; V[i] != E.end() && V[i]->u < i; ++V[i] );
   }

   inline vector<edge>::iterator begin( int u ) { return V[u]; }
   inline vector<edge>::iterator end( int u ) { return V[u+1]; }
} graph;

vector<int> discover, finish, lowlink, depth;
int Time = 0;

vector< vector<int> > children;

void dfs( int u, int dad, int d ) {
   discover[u] = lowlink[u] = Time++;
   depth[u] = d;

   for( vector<edge>::iterator it = graph.begin(u); it != graph.end(u); ++it ) {
      if( it->v == dad ) continue;

      if( discover[it->v] == -1 ) {
         dfs( it->v, u, d+1 );
         lowlink[u] <?= lowlink[it->v];
         children[u].push_back( it->v );
      } else {
         lowlink[u] <?= discover[it->v];
      }
   }

   finish[u] = Time++;
}

int is_descendant( int a, int b ) {
   return discover[b] <= discover[a] && finish[a] <= finish[b];
}

int find_related_child( int me, int descendant ) {
   int lo = 0, hi = children[me].size() - 1;
   while( lo != hi ) {
      int mid = (lo+hi) / 2;

      if( discover[descendant] > finish[ children[me][mid] ] ) lo = mid+1;
      else if( finish[descendant] < discover[ children[me][mid] ] ) hi = mid-1;
      else lo = hi = mid;
   }
   return children[me][lo];
}

int main( void ) {
   scanf( "%d%d", &n, &m );
   for( int i = 0; i < m; ++i ) {
      int u, v;
      scanf( "%d%d", &u, &v ); --u; --v;
      graph.insert_edge( edge( u, v ) );
      graph.insert_edge( edge( v, u ) );
   }
   graph.init();
   discover = finish = lowlink = depth = vector<int> (n, -1);
   children.resize( n );
   dfs( 0, -1, 0 );

   scanf( "%d", &m );
   for( int i = 0; i < m; ++i ) {
      int tip, a, b, c, d;
      scanf( "%d%d%d%d", &tip, &a, &b, &c ); --a; --b; --c;
      if( tip == 1 ) {
         scanf( "%d", &d ); --d;
         if( is_descendant( c, d ) ) swap( c, d );

         if( depth[d] != depth[c]+1 ) printf( "yes\n" );
         else if( lowlink[d] < discover[d] ) printf( "yes\n" );
         else if( is_descendant( a, d ) == is_descendant( b, d ) ) printf( "yes\n" );
         else printf( "no\n" );

      } else {
         if( !is_descendant( a, c ) && !is_descendant( b, c ) ) printf( "yes\n" );
         else if( is_descendant( a, c ) && is_descendant( b, c ) ) {
            int e = find_related_child( c, a );
            int f = find_related_child( c, b );
            if( e == f ) printf( "yes\n" );
            else if( lowlink[e] < discover[c] && lowlink[f] < discover[c] ) printf( "yes\n" );
            else printf( "no\n" );
         } else {
            if( is_descendant( a, c ) ) swap( a, b );
            int e = find_related_child( c, b );
            if( lowlink[e] < discover[c] ) printf( "yes\n" );
            else printf( "no\n" );
         }
      }
   }

   return 0;
}

File: Olympiad/COI/official/2007/policija.pas
(*
  Croatian Olympiad in Informatics 2007
  Task POLICIJA
*)

program policija;

const MAXN = 100000;
const MAXE = 500000;

type 
   pEdge = ^Edge;
   Edge = record 
      u, v : longint;
      next : pEdge;
   end;

var
   N, M, Q : longint;
   glava   : array[1..MAXN] of pEdge;
   edges   : array[1..2*MAXE] of Edge;
   djeca   : array[1..MAXN] of longint;
   it      : pEdge;
   i, j    : longint;
   Time    : longint;
   tip, a, b, c, d, e, f            : longint;
   discover, finish, lowlink, depth : array[1..MAXN] of longint;
   djecabegin, djecaend             : array[0..MAXN] of longint;

procedure swap( var a, b : longint );
var 
   tmp : longint;
begin
   tmp := a;
   a := b;
   b := tmp;
end;

procedure dfs( u, dad, d : longint );
var
   it : pEdge;
begin
   discover[u] := Time;
   lowlink[u] := Time;
   depth[u] := d;
   Time := Time + 1;

   it := glava[u];
   while it <> NIL do begin
      if it^.v <> dad then begin
         if discover[it^.v] = -1 then begin
            dfs( it^.v, u, d+1 );
            if lowlink[it^.v] < lowlink[u] then lowlink[u] := lowlink[it^.v];
         end else begin;
            if discover[it^.v] < lowlink[u] then lowlink[u] := discover[it^.v];
         end;
      end;
      it := it^.next;
   end;
   
   finish[u] := Time;
   Time := Time + 1;
end;

function is_descendant( a, b : longint ) : boolean;
begin
   is_descendant := (discover[b] <= discover[a]) and (finish[a] <= finish[b]);
end;

function find_related_child( me, descendant : longint ) : longint;
var
   lo, hi, mid : longint;
begin
   lo := djecabegin[me];
   hi := djecaend[me] - 1;
   while lo <> hi do begin
      mid := (lo+hi) div 2;
      
      if discover[descendant] > finish[ djeca[mid] ] then lo := mid+1 
      else if finish[descendant] < discover[ djeca[mid] ] then hi := mid-1
      else begin
         lo := mid;
         hi := mid;
      end;     
   end;
   find_related_child := djeca[lo];
end;

begin
   readln( N, M );
   for i := 1 to N do begin
      glava[i] := NIL;
   end;

   for i := 1 to M do begin
      j := 2*i;
      readln( edges[j].u, edges[j].v );
      edges[j].next := glava[edges[j].u];
      glava[edges[j].u] := @edges[j];

      j := j-1;
      edges[j].u := edges[j+1].v;
      edges[j].v := edges[j+1].u;
      edges[j].next := glava[edges[j].u];
      glava[edges[j].u] := @edges[j];
   end;

   for i := 1 to N do begin
      discover[i] := -1;
      finish[i] := -1;
      lowlink[i] := -1;
      depth[i] := -1;
   end;
   

   dfs( 1, 0, 0 );


   djecabegin[0] := 1;
   djecaend[0] := 1;
   for i := 1 to N do begin
      djecabegin[i] := djecaend[i-1];
      djecaend[i] := djecabegin[i];

      it := glava[i];
      while it <> NIL do begin
         if depth[it^.v] = depth[it^.u] + 1 then begin
            djeca[ djecaend[i] ] := it^.v;
            djecaend[i] := djecaend[i] + 1; 
         end;
         it := it^.next;
      end;
   end;


   readln( Q );
   for i := 1 to Q do begin
      read( tip );
      if tip = 1 then begin;
         readln( a, b, c, d );
         if is_descendant( c, d ) then swap( c, d );

         if depth[d] < depth[c]+1 then writeln( 'yes' )
         else if lowlink[d] < discover[d] then writeln( 'yes' )
         else if is_descendant( a, d ) = is_descendant( b, d ) then writeln( 'yes' )
         else writeln( 'no' );
      end else begin
         readln( a, b, c );

         if (not is_descendant( a, c )) and (not is_descendant( b, c )) then writeln( 'yes' )
         else if is_descendant( a, c ) and is_descendant( b, c ) then begin
            e := find_related_child( c, a );
            f := find_related_child( c, b );
            if e = f then writeln( 'yes' )
            else if (lowlink[e] < discover[c]) and (lowlink[f] < discover[c]) then writeln( 'yes' )
            else writeln( 'no' );
         end else begin
            if is_descendant( a, c ) then swap( a, b );
            e := find_related_child( c, b );
            if lowlink[e] < discover[c] then writeln( 'yes' )
            else writeln( 'no' );
         end;
      end;
   end;
	
end.
